lib/auditd/auditd_linux.go

```diff
@@ -0,0 +1,353 @@
+		log.WithError(err).Warn("failed to get executable name")
+		execName = UnknownValue
+	}
+
+	// Teleport never tries to get the hostname name.
+	// Let's mimic the sshd behavior.
+	const hostname = UnknownValue
+
+	return &Client{
+		execName:     execName,
+		hostname:     hostname,
+		systemUser:   msg.SystemUser,
+		teleportUser: msg.TeleportUser,
+		address:      msg.ConnAddress,
+		ttyName:      msg.TTYName,
+
+		dial: func(family int, config *netlink.Config) (NetlinkConnector, error) {
+			return netlink.Dial(family, config)
+		},
+	}
+}
+
+func getAuditStatus(conn NetlinkConnector) (*auditStatus, error) {
+	_, err := conn.Execute(netlink.Message{
+		Header: netlink.Header{
+			Type:  netlink.HeaderType(AuditGet),
+			Flags: netlink.Request | netlink.Acknowledge,
+		},
+		Data: nil,
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	msgs, err := conn.Receive()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if len(msgs) != 1 {
+		return nil, trace.Errorf("returned wrong messages number, expected 1, got: %d", len(msgs))
+	}
+
+	// auditd marshalling depends on the system architecture.
+	byteOrder := nlenc.NativeEndian()
+	status := &auditStatus{}
+
+	payload := bytes.NewReader(msgs[0].Data[:])
+	if err := binary.Read(payload, byteOrder, status); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return status, nil
+}
+
+// SendMsg sends a message. Client will create a new connection if not connected already.
+func (c *Client) SendMsg(event EventType, result ResultType) error {
+	const msgDataTmpl = `op={{ .Opcode }} acct="{{ .Msg.SystemUser }}" exe={{ .Exe }} ` +
+		`hostname={{ .Hostname }} addr={{ .Msg.ConnAddress }} terminal={{ .Msg.TTYName }} ` +
+		`{{if .Msg.TeleportUser}}teleportUser={{.Msg.TeleportUser}} {{end}}res={{ .Result }}`
+	op := eventToOp(event)
+
+	buf := &bytes.Buffer{}
+	if err := template.Must(template.New("auditd-message").
+		Parse(msgDataTmpl)).
+		Execute(buf,
+			struct {
+				Result   ResultType
+				Opcode   string
+				Exe      string
+				Hostname string
+				Msg      Message
+			}{
+				Opcode:   op,
+				Result:   result,
+				Exe:      c.execName,
+				Hostname: c.hostname,
+				Msg: Message{
+					SystemUser:   c.systemUser,
+					TeleportUser: c.teleportUser,
+					ConnAddress:  c.address,
+					TTYName:      c.ttyName,
+				},
+			}); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return trace.Wrap(c.sendMsg(netlink.HeaderType(event), buf.Bytes()))
+}
+
+func (c *Client) sendMsg(eventType netlink.HeaderType, MsgData []byte) error {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if err := c.connectUnderMutex(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	enabled, err := c.isEnabledUnderMutex()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if !enabled {
+		return ErrAuditdDisabled
+	}
+
+	msg := netlink.Message{
+		Header: netlink.Header{
+			Type:  eventType,
+			Flags: syscall.NLM_F_REQUEST | syscall.NLM_F_ACK,
+		},
+		Data: MsgData,
+	}
+
+	resp, err := c.conn.Execute(msg)
```

Comment:
Can this block or something if auditd is down or not responding? Leading to the client being deadlocked due to a mutex.